oqamaint/cli.py

```diff
@@ -0,0 +1,126 @@
+            help="Path to metadata dir (data/*.json)")
+        return parser
+
+    def _load_metadata(self):
+        path = self.options.data
+        project = {}
+
+        with open(opa.join(path, "data/repos.json"), 'r') as f:
+            target = json.load(f)
+
+        with open(opa.join(path, "data/apimap.json"), 'r') as f:
+            api = json.load(f)
+
+        with open(opa.join(path, "data/incidents.json"), 'r') as f:
+            for i, j in json.load(f).items():
+                if i.startswith('SUSE'):
+                    project[i] = SUSEUpdate(j)
+                elif i.startswith('openSUSE'):
+                    project[i] = openSUSEUpdate(j)
+                else:
+                    raise "Unknown openQA", i
+        return project, target, api
+
+    def postoptparse(self):
+        # practically quiet
+        level = logging.WARNING
+        if (self.options.debug):
+            level = logging.DEBUG
+        elif (self.options.verbose):
+            # recomended variant
+            level = logging.INFO
+
+        self.logger = logging.getLogger(self.optparser.prog)
+        self.logger.setLevel(level)
+        handler = logging.StreamHandler()
+        formatter = ColorFormatter('%(levelname)s: %(message)s')
+        handler.setFormatter(formatter)
+        self.logger.addHandler(handler)
+
+        osc.conf.get_config(override_apiurl=self.options.apiurl)
+
+        if (self.options.osc_debug):
+            osc.conf.config['debug'] = 1
+
+        self.checker = self.setup_checker()
+
+        if self.options.config:
+            self.checker.load_config(self.options.config)
+
+        if self.options.review_mode:
+            self.checker.review_mode = self.options.review_mode
+
+        if self.options.fallback_user:
+            self.checker.fallback_user = self.options.fallback_user
+
+        if self.options.fallback_group:
+            self.checker.fallback_group = self.options.fallback_group
+
+    def setup_checker(self):
+        apiurl = osc.conf.config['apiurl']
+        if not apiurl:
+            raise osc.oscerr.ConfigError("Missing apiurl")
+
+        user = self.options.user
+        group = self.options.group
+        # if no args are given, use the current oscrc "owner"
+        if user is None and group is None:
+            user = osc.conf.get_apiurl_usr(apiurl)
+
+        project, target, api = self._load_metadata()
+
+        bot = self.clazz(
+            target_repo=target,
+            project_settings=project,
+            api_map=api,
+            apiurl=apiurl,
+            dryrun=self.options.dry,
+            user=user,
+            group=group,
+            logger=self.logger)
+
+        if self.options.force:
+            bot.force = True
+
+        bot.do_comments = self.options.comment
+
+        if not self.options.openqa:
+            raise osc.oscerr.WrongArgs("missing openqa url")
+
+        bot.openqa = OpenQA_Client(server=self.options.openqa)
```

Comment:
The normal style is to initialize things like this in the bot constructor and avoid overriding `postoptparse` unless absolutely necessary. Initializing those in constructing using the arguments passed in (like the url) still allows for flexible usage, but does not inconvenience folks trying to use the class from another class as many of the bots do.

Everything in here is either copied from `ReviewBot.setup_checker` or could be placed elsewhere to match style and provide better flexibility. See `leaper.py` for a decent example. This code was previous refactored to avoid the duplication present here as all the bots had it. The standard interface is to have the constructor only require the parameters defined by `ReviewBot` and optional ones can be provided for people using class directly, but cli classes just set them afterwards to avoid duplicating the construction which has all the same stuff.

Reply:
hmm, l good idea:) only one small thing `ReviewBot.post_optparse` is buggy

Reply:
If there are issues with `ReviewBot`, lets fix them instead of copy/paste and changing the code.

Reply:
@jberry-suse how about move initialization of logger to setup_checker method? + current ReviewBot brokes without `--debug` or `--verbose` parameters:)

Reply:
Sounds good to me. Not sure what all the smiles are about. I did not write the majority of this, just advocating for not making a mess.